lib/sshutils/scp/scp.go

```diff
@@ -495,28 +507,85 @@ func (cmd *command) receiveFile(st *state, fc newFileCmd, ch io.ReadWriter) erro
 		return trace.Errorf("unexpected file copy length: %v", n)
 	}
 
-	if err := cmd.FileSystem.SetChmod(path, int(fc.Mode)); err != nil {
+	if err := cmd.FileSystem.Chmod(path, int(fc.Mode)); err != nil {
 		return trace.Wrap(err)
 	}
+	if st.stat != nil {
+		err = cmd.FileSystem.Chtimes(path, st.stat.Atime, st.stat.Mtime)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
 
-	cmd.log.Debugf("file %v(%v) copied to %v", fc.Name, fc.Length, path)
+	cmd.log.Debugf("File %v(%v) copied to %v.", fc.Name, fc.Length, path)
 	return nil
 }
 
 func (cmd *command) receiveDir(st *state, fc newFileCmd, ch io.ReadWriter) error {
+	cmd.log.Debugf("scp.receiveDir(%v): %v", cmd.Flags.Target, fc.Name)
 	targetDir := cmd.Flags.Target[0]
 
 	// copying into an existing directory? append to it:
 	if cmd.FileSystem.IsDir(targetDir) {
-		targetDir = st.makePath(targetDir, fc.Name)
-		st.push(fc.Name)
+		targetDir = st.makePath(fc.Name)
 	}
+	st.push(fc.Name, st.stat)
 
 	err := cmd.FileSystem.MkDir(targetDir, int(fc.Mode))
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	return nil
+}
+
+func (cmd *command) sendDirMode(r *reader, ch io.Writer, fileInfo FileInfo) error {
+	out := fmt.Sprintf("D%04o 0 %s\n", fileInfo.GetModePerm(), fileInfo.GetName())
+	cmd.log.WithField("cmd", out).Debug("Send directory mode.")
+	_, err := io.WriteString(ch, out)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := r.read(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (cmd *command) sendFileTimes(r *reader, ch io.Writer, fileInfo FileInfo) error {
+	out := fmt.Sprintf("T%d 0 %d 0\n",
+		fileInfo.GetModTime().Unix(),
+		fileInfo.GetAccessTime().Unix(),
+	)
+	cmd.log.WithField("cmd", out).Debug("Send file times.")
+	_, err := io.WriteString(ch, out)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(r.read())
+}
+
+func (cmd *command) sendFileMode(r *reader, ch io.Writer, fileInfo FileInfo) error {
+	out := fmt.Sprintf("C%04o %d %s\n",
+		fileInfo.GetModePerm(),
+		fileInfo.GetSize(),
+		fileInfo.GetName(),
+	)
+	cmd.log.WithField("cmd", out).Debug("Send file mode.")
+	_, err := io.WriteString(ch, out)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	return trace.Wrap(r.read())
+}
 
+func (cmd *command) updateDirTimes(path pathSegments) error {
+	if stat := path[len(path)-1].stat; stat != nil {
```

Comment:
Can `len(path)` be 0 here or is it always guaranteed to have at least 1 element?

Reply:
It is guaranteed to be non-empty [here](https://github.com/gravitational/teleport/blob/dmitri/2889-scp-p/lib/sshutils/scp/scp.go#L458-L460)